src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs

```diff
@@ -208,20 +217,74 @@ public WindowsIdentity(string sUserPrincipalName)
 
                             if (profileBuffer != null)
                                 profileBuffer.Dispose();
-
-                            _safeTokenHandle = accessTokenHandle;
                         }
                     }
                 }
             }
         }
 
-        private static SafeLsaHandle ConnectToLsa()
+        /// <summary>
+        /// To obtain SafeLsaHandle we try before for an Impersonation token that require SeTcbPrivilege, if fails we fallback
+        /// to an Identification token
+        /// </summary>
+        /// <param name="logonProcessNameBuffer"></param>
+        /// <returns></returns>
+        private static SafeLsaHandle GetLsaHandle(byte[] logonProcessNameBuffer)
         {
             SafeLsaHandle lsaHandle;
-            int ntStatus = Interop.SspiCli.LsaConnectUntrusted(out lsaHandle);
-            if (ntStatus < 0) // non-negative numbers indicate success
-                throw GetExceptionFromNtStatus(ntStatus);
+
+            // Try to get an impersonation token
+            Privilege privilege = null;
+            try
+            { 
+                // Try to enable the TCB privilege if possible
+                try
+                {
+                    privilege = new Privilege("SeTcbPrivilege");
+                    privilege.Enable();
+                }
+                catch (PrivilegeNotHeldException) { }
+
+                ushort sourceNameLength = checked((ushort)(logonProcessNameBuffer.Length));
+                using (SafeLocalAllocHandle sourceNameBuffer = Interop.Kernel32.LocalAlloc(0, new UIntPtr(sourceNameLength)))
+                {
+                    if (sourceNameBuffer.IsInvalid)
+                        throw new OutOfMemoryException();
+
+                    Marshal.Copy(logonProcessNameBuffer, 0, sourceNameBuffer.DangerousGetHandle(), logonProcessNameBuffer.Length);
+                    LSA_STRING logonProcessName = new LSA_STRING(sourceNameBuffer.DangerousGetHandle(), sourceNameLength);
+                    IntPtr dummy = IntPtr.Zero;
+                    int ntStatus = Interop.SspiCli.LsaRegisterLogonProcess(ref logonProcessName, out lsaHandle, out dummy);
+                    if (Interop.Errors.ERROR_ACCESS_DENIED == Marshal.GetLastWin32Error())
+                    {
+                        // We don't have the Tcb privilege. The best we can hope for is to get an Identification token.
+                        ntStatus = Interop.SspiCli.LsaConnectUntrusted(out lsaHandle);
+
+                        // non-negative numbers indicate success
+                        if (ntStatus < 0)
+                        { 
+                            throw GetExceptionFromNtStatus(ntStatus);
+                        }
+                    }
+                }
+            }
+            catch
```

Comment:
You can avoid the `privilege.Revert();` code duplication here by swapping the order of finally and catch. Like:

```
finally
{
    privilege?.Revert();
}
catch
{
    // protect against exception filter-based luring attacks
    throw;
}
```

Reply:
I'll wait full review to avoid too much commit.

Reply:
> You can avoid the privilege.Revert(); code duplication here by swapping the order of finally and catch

That's not valid C#, is it?

Reply:
Right, there would need to be an extra scope:

```csharp
try
{
    try
    {
         do stuff
    }
    finally 
    {
         revert
    }
}
catch
{
     // protect against exception filter-based luring attacks
    throw;
}
```

Reply:
I didn't try code above...but maybe the idea was
```cs
...
try
{
   Privilege privilege;
   try
   {
      ...
   }
   finally
   {
     privilege?.Revert();
   }
}
catch
{
   throw;
}
```